rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,476 @@
+                    unsafe {
+                        *buffer = p.as_ptr();
+                        *buffer_len = p.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
+    tx: &MQTTTransaction,
+    session_present: *mut bool,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNACK(ref ca) = msg.op {
+            unsafe {
+                *session_present = ca.session_present;
+            }
+            return 1;
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_topic(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.topic;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_message(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.message;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_subscribe_topic(tx: &MQTTTransaction,
+    i: u16,
+    buf: *mut *const u8,
+    len: *mut u32)
+    -> u8
+{
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::SUBSCRIBE(ref subv) = msg.op {
+            if (i as usize) < subv.topics.len() {
+                let topic = &subv.topics[i as usize];
+                if topic.topic_name.len() > 0 {
+                    unsafe {
+                        *len = topic.topic_name.len() as u32;
+                        *buf = topic.topic_name.as_ptr();
+                    }
+                    return 1;
+                }
+            }
+        }
+    }
+
+    unsafe {
+        *buf = ptr::null();
+        *len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_unsubscribe_topic(tx: &MQTTTransaction,
+    i: u16,
+    buf: *mut *const u8,
+    len: *mut u32)
+    -> u8
+{
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::UNSUBSCRIBE(ref unsubv) = msg.op {
```

Comment:
Can not there be more than one `UNSUBSCRIBE` in a transaction ?

Reply:
This is my last question ;-)

Reply:
Not according to the protocol, i.e. compliant brokers/clients would not create this. Forged or broken traffic might create such a situation if message IDs are used multiple times for separate UNSUBSCRIBE events.
I think I might really need the approach from https://github.com/OISF/suricata/pull/5208#discussion_r459466182 to make sure we get all of their topics if we want this function to work in this case as well (which I think we do!)

Reply:
See https://github.com/satta/suricata/commit/058c91b79cd43e587a49670eea0bd5e4580aa2ee

Reply:
This commit looks good to me

Reply:
Cool, thanks for having a look. I will revisit and review the transaction code again to better define how the message ID ("packet identifier" in the spec) is used to group messages into transactions, in particular regarding ID reuse as addressed in the spec:

> The Packet Identifier becomes available for reuse after the Client has processed the corresponding acknowledgement packet. In the case of a QoS 1 PUBLISH this is the corresponding PUBACK; in the case of QoS 2 it is PUBCOMP. For SUBSCRIBE or UNSUBSCRIBE it is the corresponding SUBACK or UNSUBACK.

This means that we should make sure to never add to a 'completed' transaction where an acknowledgement has been seen. I'll make sure that is the case.